common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -586,6 +592,7 @@ public void onData(String streamId, ByteBuffer buf) throws IOException {
             deferredBufs = null;
             return;
           }
+          abortIfNecessary();
```

Comment:
I don't think we should invoke this here.
Invoking inside `onComplete` should be sufficient.

Reply:
Purpose is to only do this check once per block instead of once per buf.

Reply:
@Victsm This change is to address this comment:
https://github.com/apache/spark/pull/30433#discussion_r538947833

The scenario is that there can be pending streams which are waiting on the lock for the `partitionInfo` and meanwhile the exception threshold has met. When the pending stream acquires the lock it will attempt to write to the data file even though the exception threshold is reached. 

I have added the UT `testPendingBlockIsAbortedImmediately` to verify this. 

> Purpose is to only do this check once per block instead of once per buf and avoid throwing unnecessary exceptions from onData which leads to channel close.

We already throw IOExceptions when write to data file fails. I don't see how throwing `IOException exceeded threshold` makes it any worse.

Reply:
The way I see the issue mentioned in that comment is that we are not preventing new blocks to be merged when the IOException threshold is reached.
To do that, we only need to invoke `abortIfNecessary` inside `onComplete`, whether we still have any deferredBuf to write at that point.
This way, for normal case without IOException, we are only invoking `abortIfNecessary` once per block.
By invoking it here, we would invoke it once per buf for normal case.

Of course, if we only check inside `onComplete`, we would delay rejection of these pending blocks until we reach their stream's end.
I think this is a reasonable tradeoff to make, considering that majority of the time the code is executing for normal case instead of the exception case.

Reply:
hmmm. I think the intention is to not have the server attempt writing if the threshold is reached for the partition. Probably this check here makes this behavior more accurate.

However, I don't have a strong opinion on this since the assumption is that if these number of IOExceptions have already reached the threshold, any further writes will result in an IOException as well. With that assumption, the write in `onData` after this threshold is met  will very likely throw `IOException` as well and since the  threshold is already met, the server will instead throw `IOExceptions exceeded threshold`.

I can remove it from `onData`. @Ngone51 What do you think?

Reply:
It's OK if the further writes throw IOExceptions. I just worry, can we handle it correctly if writes succeed? Or it may lead to some undefined situation which could avoid later shuffle fetch in turn.

So I'd like to avoid any writes in the first place (`onData`) so we ensure that we don't go into any undefined situation. And I think, maybe, we can update it a little, like this - we still receive data in `onData` but just not write when the partition already reach IO threshold and only call `abortIfNecessary` inside `onCompelte`.

WDYT? @Victsm @otterc

Reply:
Even If the write succeeds, I don't see an issue. If there are no further failures with this block it will be merged successfully. If there are more failures later, then they will be handled in the same way as now.  I could add UTs to verify this as well.

> we still receive data in onData but just not write when the partition already reach IO threshold and only call abortIfNecessary inside `onComplete`.

My thought about this is that if a block is completing and doesn't have any deferred bufs then we should let it complete without failures if possible. 

Again, I don't have a strong opinion about this. IMO reaching the threshold is indicative of a persistent issue and any further writes will faill.

Reply:
I agree with @Ngone51 - if we failed on a previous write, there should not be any subsequent writes: it will lead to inconsistent state (for example, if previous write failed due to lack of space and subsequent write can go through due to some async cleanup).
It is better to avoid these states rather than reason about them based on current code state - as things evolve, it makes it more brittle.

Reply:
If we want consistency, i.e., no more write after reaching IOException threshold, then we should alway invoke `abortIfNecessary` in `onComplete` to check, and if that's what we want, only invoking `abortIfNecessary` in `onComplete` should be sufficient.

Reply:
@Ngone51 @mridulm , by throwing exception inside `onComplete`, we are still effectively rejecting any new blocks from being successfully appended to the file.

Reply:
Had an offline discussion with @mridulm and @otterc , the concern about closing channels with throwing exception mid-stream seems negligible since it only happens after reaching the max IOException count.
Calling `abortIfNecessary` inside `onData` should also have negligible performance implications.
With those, I think it should be fine to keep this part of the code as is.

Reply:
I'm wondering does the server closes the channel or client would stop streaming the remaining data when we throw the exception in `onData`. Otherwise, we'd receive the following data from the client and throw the exception for multiple times.

(I didn't find anywhere we close the channel or the client may have a chance to stop streaming. I may miss it somewhere.)

Reply:
The client does receive the exception thrown from `onData`. I simulated an exception from `onData` at the server for a particular push block.
Below are the logs of the client. 
Note: I am running an older version of magnet which still uses `ShuffleBlockId` for shuffle push and some other classes are old.

```
20/12/11 19:14:48 INFO TransportClientFactory: Successfully created connection to ltx1-hcl3213.grid.linkedin.com/10.150.24.33:7337 after 12 ms (5 ms spent in bootstraps)
20/12/11 19:14:56 ERROR RetryingBlockFetcher: Failed to fetch block shuffle_1_7_7, and will not retry (1 retries)
org.apache.spark.network.shuffle.BlockPushException: ^H^@^@^@^_application_1602506816280_53624^@^@^@
shuffle_1_7_7^@^@^@^Gjava.io.IOException: Destination failed while reading stream
        at org.apache.spark.network.server.TransportRequestHandler$3.onFailure(TransportRequestHandler.java:244)
        at org.apache.spark.network.client.StreamInterceptor.exceptionCaught(StreamInterceptor.java:58)
        at org.apache.spark.network.util.TransportFrameDecoder.exceptionCaught(TransportFrameDecoder.java:188)
        at org.spark_project.io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
        at org.spark_project.io.netty.channel.AbstractChannelHandlerContext.notifyHandlerException(AbstractChannelHandlerContext.java:850)
        at org.spark_project.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:364)
        at org.spark_project.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at org.spark_project.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at org.spark_project.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359)
        at org.spark_project.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at org.spark_project.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at org.spark_project.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935)
        at org.spark_project.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:138)
        at org.spark_project.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
        at org.spark_project.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)
        at org.spark_project.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)
        at org.spark_project.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
        at org.spark_project.io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at org.spark_project.io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: FAILING this stream
```

After this, the client logs show that the server has terminated the connection.
```
20/12/11 19:14:57 WARN TransportChannelHandler: Exception in connection from ltx1-hcl3412.grid.linkedin.com/10.150.55.78:7337
java.io.IOException: Connection reset by peer
        at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
        at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
        at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
        at sun.nio.ch.IOUtil.read(IOUtil.java:192)
        at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380)
        at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288)
        at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1106)
        at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:343)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:748)
20/12/11 19:14:57 ERROR TransportResponseHandler: Still have 38 requests outstanding when connection from ltx1-hcl3412.grid.linkedin.com/10.150.55.78:7337 is closed
```